skip mismatched webhookAccessor and object

gongguan opened the pull request:
/kind cleanup

**Special notes for your reviewer**:
Skip building WebhookInvocation when WebhookAccessor and object mismatch, since it's expensive to build WebhookInvocation.

Before
```
BenchmarkShouldCallHookWithComplexSelector-4              942778              1214 ns/op             288 B/op          6 allocs/op
BenchmarkShouldCallHookWithComplexRule-4                   63638             20492 ns/op              64 B/op          1 allocs/op
BenchmarkShouldCallHookWithComplexSelectorAndRule-4        61670             20800 ns/op              65 B/op          1 allocs/op

```
After
```
BenchmarkShouldCallHookWithComplexSelector-4             6707960               218 ns/op              32 B/op          2 allocs/op
BenchmarkShouldCallHookWithComplexRule-4                 5192454               216 ns/op              32 B/op          2 allocs/op
BenchmarkShouldCallHookWithComplexSelectorAndRule-4      5837241               221 ns/op              32 B/op          2 allocs/op
```

```release-note
None
```

gongguan commented:
Changed test case because panic will be thrown if NamespaceSelector is nil.
While we can ensure ValidatingWebhook 's NamespaceSelector always not nil, right? Otherwise we can also produce a panic when removing this line https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook_test.go#L114

gongguan commented:
/assign @liggitt

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -142,6 +142,17 @@ func (a *Webhook) ValidateInitialization() error {
 func (a *Webhook) ShouldCallHook(h webhook.WebhookAccessor, attr admission.Attributes, o admission.ObjectInterfaces) (*WebhookInvocation, *apierrors.StatusError) {
 	var err *apierrors.StatusError
+
+	matches, err := a.namespaceMatcher.MatchNamespaceSelector(h, attr)
+	if !matches || err != nil {
+		return nil, err
+	}
```
This puts a namespace lookup (potentially a live network request) and several error paths ahead of checking if the webhook's rules even apply to this request. I'd like to see benchmark evidence that this is actually cheaper, and consider how to avoid returning an error if the webhook's rules don't even apply to the request.

gongguan replied:
Updated, result:
BenchmarkShouldCallHook-4         329312              3182 ns/op
BenchmarkShouldCallHook-4         510973              2305 ns/op

gongguan replied:
> how to avoid returning an error if the webhook's rules don't even apply to the request.

I have a question: where is the error happened, when we build invocation WebhookAccessor and Attributes won't change, right?

gongguan replied:
Updated benchmark, make NewValidatingWebhookAccessor outside loop, result is
BenchmarkShouldCallHook-4        1329668               902 ns/op             288 B/op          6 allocs/op
BenchmarkShouldCallHook-4        8379460               140 ns/op              32 B/op          2 allocs/op

@liggitt  Do you think this bench test reasonable?

gongguan replied:
> consider how to avoid returning an error if the webhook's rules don't even apply to the request.

Do you mean MatchNamespaceSelector returns an error while invocation(rules don't even apply to the request) is nil? We want to continue next hook but returned an error. Yeah, I haven't caught this scenario.

This just hide the error(mostly a get request, other occurred error might be logic bug) when no rules applied to request, once rules applied, error would happen, right?

liggitt replied:
> Do you mean MatchNamespaceSelector returns an error while invocation(rules don't even apply to the request) is nil? We want to continue next hook but returned an error. Yeah, I haven't caught this scenario.

Exactly. We should not return an error here for webhooks which do not even apply to the incoming request.

liggitt replied:
> This puts a namespace lookup (potentially a live network request) ... ahead of checking if the webhook's rules even apply to this request.

this is still outstanding as well

gongguan replied:
Yeah, it can't avoid returning an error if the webhook's rules don't even apply to the request, so I close it.

gongguan replied:
> Exactly. We should not return an error here for webhooks which do not even apply to the incoming request.

@liggitt addressed it. PTAL.

gongguan replied:
Sorry, it's not right. it will break https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go#L175

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -142,6 +142,17 @@ func (a *Webhook) ValidateInitialization() error {
+
+	matches, err = a.objectMatcher.MatchObjectSelector(h, attr)
+	if !matches || err != nil {
+		return nil, err
```
same comment about not returning an error early here before checking if the webhook even applies to this request

gongguan committed:
skip mismatched webhookAccessor and object

gongguan commented:
/retest

gongguan commented:
Close it because this can't avoid returning an error if the webhook's rules don't even apply to the request.

gongguan commented:
address @liggitt https://github.com/kubernetes/kubernetes/pull/92696#discussion_r451012290 and reopen it

gongguan commented:
/retest

gongguan commented:
pod pending timeout

/test pull-kubernetes-integration

gongguan commented:
> Exactly. We should not return an error here for webhooks which do not even apply to the incoming request.

@liggitt I have addressed your comment and reopen this pr, PTAL.

liggitt commented:
in determining if this is a beneficial change, benchmark-wise, there are a few scenarios we should consider:

* webhook config with complex selector (namespace or object or both) and simple matching rule
* webhook config with simple or absent selector and complex matching rules
* webhook config with complex selector (namespace or object or both) and complex matching rules

Comparing the before/after for all three scenarios will help determine whether there are cases where this change actually decreases performance

gongguan commented:
> in determining if this is a beneficial change, benchmark-wise, there are a few scenarios we should consider:

Do you mean I should add at least three benchmark tests to cover these scenarios ?

liggitt commented:
> @liggitt Do you mean I should add at least three benchmark tests to cover these scenarios ?

yes, extend the current benchmark to run those three scenarios

gongguan committed:
extend ShouldCallHook benchmark to verify performance imporvement

gongguan commented:
@liggitt I have extended current benchmark tests, the improvement is more obvious:

Before
```
BenchmarkShouldCallHookWithComplexSelector-4              942778              1214 ns/op             288 B/op          6 allocs/op
BenchmarkShouldCallHookWithComplexRule-4                   63638             20492 ns/op              64 B/op          1 allocs/op
BenchmarkShouldCallHookWithComplexSelectorAndRule-4        61670             20800 ns/op              65 B/op          1 allocs/op

```
After
```
BenchmarkShouldCallHookWithComplexSelector-4             6707960               218 ns/op              32 B/op          2 allocs/op
BenchmarkShouldCallHookWithComplexRule-4                 5192454               216 ns/op              32 B/op          2 allocs/op
BenchmarkShouldCallHookWithComplexSelectorAndRule-4      5837241               221 ns/op              32 B/op          2 allocs/op
```

liggitt commented:
/lgtm
/approve

it makes sense that a complex selector is faster at short-circuiting than a complex list of rules

gongguan commented:
Thanks for reviewing @liggitt 

/sig scalability
/cc @wojtek-t

gongguan commented:
/test pull-kubernetes-kubemark-e2e-gce-big

wojtek-t commented:
nice!

justaugustus commented:
/test pull-kubernetes-e2e-kind-ipv6
